### DISCUSSION TITLE
[Help] A callback to know a workflow is completed?

### DISCUSSION THREAD

**Discussion by lf-jason-jia** [elsa-workflows/elsa-core#3977] [Q&A] [answered]
Hi I am looking into the StartAt activity. I know that we could know that an activity is about to finish in its OnExecute or OnResume (for time-based activity). But is there a general way to know a completion of an activity without adding code in an activity's OnExecute or OnResume? I guess you could probably listen to some event?

Thanks!

**Comment by sfmskywalker** [comment, marked as answer]
Sure. For Elsa 2, you can handle the `ActivityExecuted` notification.
For Elsa 3, there is currently no event that is similar, but we should probably add it. What you can do, is install custom activity execution middleware that gets executed for each activity. We'll probably provide a middleware component that will publish notifications like ActivityExecuting and ActivityExecuted.

**Reply by lf-jason-jia** [reply]
Thank you for the help. I am able to implement an `INotificationHandler<ActivityExecuted>` and register it using `AddNotificationHandler`.
